<a href="https://colab.research.google.com/github/ignaciosoto04/Projects/blob/main/Credit_Score_Classification_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import Libraries

In [87]:
import numpy as np
import pandas as pd

Read the data

In [88]:
credit_info = pd.read_csv('CreditInformation.csv')
credit_info.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (26) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


ID Customer_ID     Month           Name   Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2  0x1604   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3  0x1605   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4  0x1606   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  ...  Credit_Mix  \
0      19114.12            1824.843333                  3  ...           _   
1      19114.12                    NaN                  3  ...        Good   
2      19114.12                    NaN                  3  ...        Good   
3      19114.12                    NaN                  3  ...        Good   
4      19114.12            1824.843333                  3  ...        Good   

   Outstanding_Debt Credit_Utilization_Ratio     Credit_History_Age  \
0            809.98                26.822620  22 Years and 1 Months   
1            809.98                31.944960                    NaN   
2            809.98                28.609352  22 Years and 3 Months   
3            809.98                31.377862  22 Years and 4 Months   
4            809.98                24.797347  22 Years and 5 Months   

   Payment_of_Min_Amount Total_EMI_per_month Amount_invested_monthly  \
0                     No           49.574949             80.41529544   
1                     No           49.574949             118.2802216   
2                     No           49.574949             81.69952126   
3                     No           49.574949             199.4580744   
4                     No           49.574949             41.42015309   

                  Payment_Behaviour Monthly_Balance Credit_Score  
0   High_spent_Small_value_payments     312.4940887         Good  
1    Low_spent_Large_value_payments     284.6291625         Good  
2   Low_spent_Medium_value_payments     331.2098629         Good  
3    Low_spent_Small_value_payments     223.4513097         Good  
4  High_spent_Medium_value_payments      341.489231         Good  

[5 rows x 28 columns]

Check columns

In [89]:
credit_info.columns

Index(['ID', 'Customer_ID', 'Month', 'Name', 'Age', 'SSN', 'Occupation',
       'Annual_Income', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts',
       'Num_Credit_Card', 'Interest_Rate', 'Num_of_Loan', 'Type_of_Loan',
       'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Mix', 'Outstanding_Debt',
       'Credit_Utilization_Ratio', 'Credit_History_Age',
       'Payment_of_Min_Amount', 'Total_EMI_per_month',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')

Delete unnecesary columns

In [90]:
credit_info.drop(columns=['Name', 'SSN'], inplace = True)

Set ID as index

In [91]:
credit_info.set_index('ID', inplace = True)

Concise summary of the dataframe

In [92]:
credit_info.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 0x1602 to 0x25fed
Data columns (total 25 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Customer_ID               100000 non-null  object 
 1   Month                     100000 non-null  object 
 2   Age                       100000 non-null  object 
 3   Occupation                100000 non-null  object 
 4   Annual_Income             100000 non-null  object 
 5   Monthly_Inhand_Salary     84998 non-null   float64
 6   Num_Bank_Accounts         100000 non-null  int64  
 7   Num_Credit_Card           100000 non-null  int64  
 8   Interest_Rate             100000 non-null  int64  
 9   Num_of_Loan               100000 non-null  object 
 10  Type_of_Loan              88592 non-null   object 
 11  Delay_from_due_date       100000 non-null  int64  
 12  Num_of_Delayed_Payment    92998 non-null   object 
 13  Changed_Credit_Limit      100000 non-null  

Unique values

In [93]:
for i in credit_info:
  print({i})
  print(credit_info[i].unique())
  print()

{'Customer_ID'}
['CUS_0xd40' 'CUS_0x21b1' 'CUS_0x2dbc' ... 'CUS_0xaf61' 'CUS_0x8600'
 'CUS_0x942c']

{'Month'}
['January' 'February' 'March' 'April' 'May' 'June' 'July' 'August']

{'Age'}
['23' '-500' '28_' ... '4808_' '2263' '1342']

{'Occupation'}
['Scientist' '_______' 'Teacher' 'Engineer' 'Entrepreneur' 'Developer'
 'Lawyer' 'Media_Manager' 'Doctor' 'Journalist' 'Manager' 'Accountant'
 'Musician' 'Mechanic' 'Writer' 'Architect']

{'Annual_Income'}
['19114.12' '34847.84' '34847.84_' ... '20002.88' '39628.99' '39628.99_']

{'Monthly_Inhand_Salary'}
[1824.843333         nan 3037.986667 ... 3097.008333 1929.906667
 3359.415833]

{'Num_Bank_Accounts'}
[   3    2    1    7    4    0    8    5    6    9   10 1414 1231   67
  572 1488   91  528 1647 1696 1338  649  889 1668  685  857  975 1496
 1534 1620   37 1388 1429 1332 1588  120 1777 1096  803  494  744 1139
  831 1613  741  121  665 1748 1644  823 1356 1651   87  711  450  210
 1671 1722  648  672 1662 1495  510  666  429 1131 1434  

I am going to fix all the issues for each column.

Customer_ID and Month are ok.

Next column is Age. It has errors.

# Age

In [94]:
credit_info['Age'].describe()

count     100000
unique      1788
top           38
freq        2833
Name: Age, dtype: object

Before converting it to float, first I have to fix the values that have "_"

In [95]:
credit_info['Age'] = credit_info['Age'].str.strip('_')

In [96]:
credit_info['Age'] = credit_info['Age'].astype('float')

Now, I am only consider the ages between 18 and 65, since I assume that these are the ages where people work

In [97]:
credit_info['Age'].describe()

count    100000.000000
mean        110.649700
std         686.244717
min        -500.000000
25%          24.000000
50%          33.000000
75%          42.000000
max        8698.000000
Name: Age, dtype: float64

In [98]:
# List1 is created to find the ages not between 18 and 65 inclusive
list1 = []
for i in credit_info['Age']:
  if i < 18 or i > 65:
    list1.append(i)
print(list1)

[-500.0, 7580.0, -500.0, 181.0, 995.0, 5079.0, 6409.0, -500.0, 16.0, 16.0, 16.0, 16.0, 16.0, 16.0, 7080.0, 849.0, -500.0, 3885.0, 15.0, 15.0, 15.0, 15.0, 16.0, 16.0, 16.0, 16.0, 3052.0, 14.0, 14.0, 14.0, 14.0, 14.0, 14.0, 14.0, 14.0, 5342.0, 16.0, 16.0, 16.0, 16.0, 16.0, 16.0, 16.0, 16.0, 17.0, 17.0, 17.0, 4431.0, -500.0, 2657.0, 2111.0, 1032.0, 15.0, 15.0, 16.0, 16.0, 16.0, 16.0, 16.0, -500.0, 456.0, 5717.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, -500.0, -500.0, -500.0, 3169.0, 1191.0, -500.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 7140.0, -500.0, 14.0, 14.0, 14.0, 15.0, 15.0, 15.0, 15.0, 15.0, -500.0, 8233.0, 6955.0, -500.0, 2534.0, 3115.0, 7657.0, 5112.0, -500.0, 6452.0, 15.0, 15.0, 16.0, 16.0, 16.0, 16.0, 16.0, 16.0, 2744.0, -500.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 17.0, 1439.0, 5795.0, 4872.0, -500.0, 1772.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 15.0, 1383.0, 5657.0, -500.0, 1934.0, -500.0, 8352.0, -500.0, 3

Ages not between 18 and 65, will be replaced by 0

In [99]:
for i in credit_info['Age']:
  if i in list1:
    credit_info['Age'] = credit_info['Age'].replace(i, 0)

In [100]:
credit_info['Age'].describe()

count    100000.000000
mean         31.504640
std          13.625486
min           0.000000
25%          24.000000
50%          32.000000
75%          41.000000
max          56.000000
Name: Age, dtype: float64

In [101]:
#No Na
credit_info['Age'].isna().sum()

0

Now, I will replace the 0s with the maximum age for each Customer_ID

In [102]:
credit_info['Age'] = credit_info.groupby(by=['Customer_ID'])['Age'].transform('max')

In [103]:
credit_info[['Customer_ID','Age']].head(16)

Customer_ID   Age
ID                      
0x1602   CUS_0xd40  23.0
0x1603   CUS_0xd40  23.0
0x1604   CUS_0xd40  23.0
0x1605   CUS_0xd40  23.0
0x1606   CUS_0xd40  23.0
0x1607   CUS_0xd40  23.0
0x1608   CUS_0xd40  23.0
0x1609   CUS_0xd40  23.0
0x160e  CUS_0x21b1  28.0
0x160f  CUS_0x21b1  28.0
0x1610  CUS_0x21b1  28.0
0x1611  CUS_0x21b1  28.0
0x1612  CUS_0x21b1  28.0
0x1613  CUS_0x21b1  28.0
0x1614  CUS_0x21b1  28.0
0x1615  CUS_0x21b1  28.0

Check if Age is fixed

In [104]:
print("AGE:")
print("Number of NAs = ",credit_info['Age'].isna().value_counts())
print("Number of Nulls = ",credit_info['Age'].isnull().value_counts())
print("Number of 0s = ",credit_info['Age'][credit_info['Age']==0].value_counts())

AGE:
Number of NAs =  False    100000
Name: Age, dtype: int64
Number of Nulls =  False    100000
Name: Age, dtype: int64
Number of 0s =  0.0    5504
Name: Age, dtype: int64


Not fixed yet, there are 5504 values that are 0.

In [105]:
credit_info[credit_info['Customer_ID']=='CUS_0x4080']

Customer_ID     Month  Age Occupation Annual_Income  \
ID                                                           
0x17b2  CUS_0x4080   January  0.0   Mechanic      29469.98   
0x17b3  CUS_0x4080  February  0.0   Mechanic      29469.98   
0x17b4  CUS_0x4080     March  0.0   Mechanic      29469.98   
0x17b5  CUS_0x4080     April  0.0   Mechanic      29469.98   
0x17b6  CUS_0x4080       May  0.0   Mechanic      29469.98   
0x17b7  CUS_0x4080      June  0.0   Mechanic     29469.98_   
0x17b8  CUS_0x4080      July  0.0   Mechanic      29469.98   
0x17b9  CUS_0x4080    August  0.0   Mechanic      29469.98   

        Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
ID                                                                  
0x17b2            2227.831667               1231                7   
0x17b3            2227.831667                  7                7   
0x17b4            2227.831667                  7                7   
0x17b5            2227.831667                  7                7   
0x17b6            2227.831667                  7                7   
0x17b7                    NaN                  7                7   
0x17b8                    NaN                  7                7   
0x17b9            2227.831667                  7                7   

        Interest_Rate Num_of_Loan  ... Credit_Mix  Outstanding_Debt  \
ID                                 ...                                
0x17b2             24           5  ...          _           3421.66   
0x17b3             24           5  ...        Bad           3421.66   
0x17b4             24        -100  ...          _           3421.66   
0x17b5             24           5  ...        Bad           3421.66   
0x17b6             24           5  ...        Bad           3421.66   
0x17b7             24           5  ...        Bad           3421.66   
0x17b8             24           5  ...        Bad           3421.66   
0x17b9             24           5  ...        Bad           3421.66   

       Credit_Utilization_Ratio     Credit_History_Age  Payment_of_Min_Amount  \
ID                                                                              
0x17b2                24.639658                    NaN                    Yes   
0x17b3                30.268411  13 Years and 3 Months                     NM   
0x17b4                31.212269  13 Years and 4 Months                    Yes   
0x17b5                33.298654  13 Years and 5 Months                    Yes   
0x17b6                38.177311  13 Years and 6 Months                    Yes   
0x17b7                30.986998  13 Years and 7 Months                    Yes   
0x17b8                33.283061  13 Years and 8 Months                    Yes   
0x17b9                32.962950  13 Years and 9 Months                    Yes   

       Total_EMI_per_month Amount_invested_monthly  \
ID                                                   
0x17b2           69.685459             59.82559612   
0x17b3           69.685459             125.4917683   
0x17b4           69.685459             32.70665182   
0x17b5           69.685459             27.25079463   
0x17b6           69.685459             71.26438577   
0x17b7           69.685459             117.3103161   
0x17b8           69.685459             115.4534157   
0x17b9           69.685459             24.06613145   

                       Payment_Behaviour Monthly_Balance Credit_Score  
ID                                                                     
0x17b2                            !@9#%8      363.272112     Standard  
0x17b3   High_spent_Small_value_payments     287.6059398     Standard  
0x17b4   High_spent_Large_value_payments     360.3910563     Standard  
0x17b5  High_spent_Medium_value_payments     375.8469135         Poor  
0x17b6   Low_spent_Medium_value_payments     361.8333223         Poor  
0x17b7    Low_spent_Small_value_payments      325.787392         Poor  
0x17b8    Low_spent_Small_value_payments     327.6442924

In [106]:
print(credit_info[['Customer_ID']][credit_info['Age'] == 0].count())
print()
credit_info[['Customer_ID']][credit_info['Age'] == 0].groupby('Customer_ID').Customer_ID.count()

Customer_ID    5504
dtype: int64



Customer_ID
CUS_0x1018    8
CUS_0x1041    8
CUS_0x107c    8
CUS_0x107e    8
CUS_0x1123    8
             ..
CUS_0xf60     8
CUS_0xf95     8
CUS_0xfaa     8
CUS_0xfae     8
CUS_0xfdd     8
Name: Customer_ID, Length: 688, dtype: int64

This is because there was no Age in those Customers

Two solutions: delete them or assign them the average or median Age

In [107]:
import statistics

In [108]:
print("AGE:")
print("Average =", round(statistics.mean(credit_info['Age']),0))
print("Median =", statistics.median(credit_info['Age']))

AGE:
Average = 33.0
Median = 33.0


Solution 1:

Average and Median are the same

I will change the 0s for the average

In [109]:
#for i in credit_info['Age']:
#  if i == 0:
#    credit_info['Age'] = credit_info['Age'].replace(i, round(statistics.mean(credit_info['Age']),0))

Solution 2:

Delete all customers where Age is 0

In [110]:
#credit_info.drop(credit_info[credit_info['Age']==0].index, inplace=True) 

All good for Age.

# Occupation

Reviewing values

In [111]:
credit_info['Occupation'].unique()

array(['Scientist', '_______', 'Teacher', 'Engineer', 'Entrepreneur',
       'Developer', 'Lawyer', 'Media_Manager', 'Doctor', 'Journalist',
       'Manager', 'Accountant', 'Musician', 'Mechanic', 'Writer',
       'Architect'], dtype=object)

2 problems found:


1.   Value '_______'
2.   Value '_' in Media_Manager



First, delete de '_' in Media Manager

In [112]:
for i in credit_info['Occupation']:
  if i == 'Media_Manager':
    credit_info['Occupation'] = credit_info['Occupation'].replace(i, 'Media Manager')

In [113]:
credit_info['Occupation'].unique()

array(['Scientist', '_______', 'Teacher', 'Engineer', 'Entrepreneur',
       'Developer', 'Lawyer', 'Media Manager', 'Doctor', 'Journalist',
       'Manager', 'Accountant', 'Musician', 'Mechanic', 'Writer',
       'Architect'], dtype=object)

Second, fix the '_____' value

1. Change to Na

In [114]:
for i in credit_info['Occupation']:
  if i == '_______':
    credit_info['Occupation'] = credit_info['Occupation'].replace(i, np.nan)

In [115]:
credit_info['Occupation'].unique()

array(['Scientist', nan, 'Teacher', 'Engineer', 'Entrepreneur',
       'Developer', 'Lawyer', 'Media Manager', 'Doctor', 'Journalist',
       'Manager', 'Accountant', 'Musician', 'Mechanic', 'Writer',
       'Architect'], dtype=object)

2. Fill Occupations NA's with the mode

In [116]:
from scipy import stats

In [117]:
credit_info['Occupation'] = credit_info.groupby('Customer_ID')['Occupation'].transform(lambda x: x.fillna(stats.mode(x)[0][0]))

In [118]:
credit_info['Occupation'].unique()

array(['Scientist', 'Teacher', 'Engineer', 'Entrepreneur', 'Developer',
       'Lawyer', 'Media Manager', 'Doctor', 'Journalist', 'Manager',
       'Accountant', 'Musician', 'Mechanic', 'Writer', 'Architect'],
      dtype=object)

All done with Occupation

# Annual Income

In [119]:
credit_info['Annual_Income'].unique()

array(['19114.12', '34847.84', '34847.84_', ..., '20002.88', '39628.99',
       '39628.99_'], dtype=object)

Erase '_' in some values

In [120]:
credit_info['Annual_Income'] = credit_info['Annual_Income'].str.strip('_')

Change it to float

In [121]:
credit_info['Annual_Income'] = credit_info['Annual_Income'].astype('float')

In [122]:
credit_info['Annual_Income'].unique()

array([ 19114.12,  34847.84, 143162.64, ...,  37188.1 ,  20002.88,
        39628.99])

In [123]:
credit_info['Annual_Income'].describe()

count    1.000000e+05
mean     1.764157e+05
std      1.429618e+06
min      7.005930e+03
25%      1.945750e+04
50%      3.757861e+04
75%      7.279092e+04
max      2.419806e+07
Name: Annual_Income, dtype: float64

All done with Annual Income

# Monthly Inhand Salary

In [124]:
credit_info['Monthly_Inhand_Salary'].describe()

count    84998.000000
mean      4194.170850
std       3183.686167
min        303.645417
25%       1625.568229
50%       3093.745000
75%       5957.448333
max      15204.633330
Name: Monthly_Inhand_Salary, dtype: float64

It has less than 100000 values, so there might be some Na's

In [125]:
credit_info['Monthly_Inhand_Salary'].unique()

array([1824.843333,         nan, 3037.986667, ..., 3097.008333,
       1929.906667, 3359.415833])

In [126]:
credit_info['Monthly_Inhand_Salary'].isna().sum()

15002

Check min, max, median, mean for each customer to see which is the best stats to change instead of the Na's

In [127]:
credit_info.groupby(['Customer_ID'])['Monthly_Inhand_Salary'].agg(['min','max','median','mean'])

min          max       median         mean
Customer_ID                                                    
CUS_0x1000   2706.161667  2706.161667  2706.161667  2706.161667
CUS_0x1009   4250.390000  4250.390000  4250.390000  4250.390000
CUS_0x100b   9549.782500  9549.782500  9549.782500  9549.782500
CUS_0x1011   5208.872500  5208.872500  5208.872500  5208.872500
CUS_0x1013   7962.415000  7962.415000  7962.415000  7962.415000
...                  ...          ...          ...          ...
CUS_0xff3    1176.398750  1176.398750  1176.398750  1176.398750
CUS_0xff4    2415.855000  2415.855000  2415.855000  2415.855000
CUS_0xff6    9727.326667  9727.326667  9727.326667  9727.326667
CUS_0xffc    5218.097500  5218.097500  5218.097500  5218.097500
CUS_0xffd    3749.870000  3749.870000  3749.870000  3749.870000

[12500 rows x 4 columns]

It seems that all are the same, so I will change the Nas for the Mode

In [128]:
credit_info['Monthly_Inhand_Salary'] = credit_info.groupby('Customer_ID')['Monthly_Inhand_Salary'].transform(lambda x: x.fillna(stats.mode(x)[0][0]))

In [129]:
credit_info['Monthly_Inhand_Salary'].describe()

count    99740.000000
mean      4198.094667
std       3186.506764
min        303.645417
25%       1626.047500
50%       3097.008333
75%       5957.715000
max      15204.633330
Name: Monthly_Inhand_Salary, dtype: float64

All done with Monthly Inhand Salary

# Num Bank Accounts

Check values

In [130]:
credit_info['Num_Bank_Accounts'].describe()

count    100000.000000
mean         17.091280
std         117.404834
min          -1.000000
25%           3.000000
50%           6.000000
75%           7.000000
max        1798.000000
Name: Num_Bank_Accounts, dtype: float64

There are values that have to be fixed: negatives and thousands

Check which stat method use to change those wrong values

In [131]:
from statistics import mode

In [132]:
credit_info.groupby(['Customer_ID'])['Num_Bank_Accounts'].agg(['min','max','median','mean',pd.Series.mode])

min  max  median  mean mode
Customer_ID                             
CUS_0x1000     6    6     6.0   6.0    6
CUS_0x1009     6    6     6.0   6.0    6
CUS_0x100b     1    1     1.0   1.0    1
CUS_0x1011     3    3     3.0   3.0    3
CUS_0x1013     3    3     3.0   3.0    3
...          ...  ...     ...   ...  ...
CUS_0xff3      0    0     0.0   0.0    0
CUS_0xff4      8    8     8.0   8.0    8
CUS_0xff6      5    5     5.0   5.0    5
CUS_0xffc      6    6     6.0   6.0    6
CUS_0xffd      8    8     8.0   8.0    8

[12500 rows x 5 columns]

Can't use min because there are negatives values.

Can't use max because the over thousands values.

Median, Mean and Mode seem to be the same. I will use Median in this case.

In [133]:
credit_info['Num_Bank_Accounts'] = credit_info.groupby(by=['Customer_ID'])['Num_Bank_Accounts'].transform('median')

Check if issues were solved

In [134]:
credit_info['Num_Bank_Accounts'].describe()

count    100000.000000
mean          5.369600
std           2.592987
min          -1.000000
25%           3.000000
50%           5.000000
75%           7.000000
max          10.500000
Name: Num_Bank_Accounts, dtype: float64

Thousand values were corrected.

Now, I will fix the negative ones.

Change negatives for 0s because there are not negatives number of accounts

In [135]:
credit_info[['Customer_ID','Num_Bank_Accounts']][credit_info['Num_Bank_Accounts'] < 0]

Customer_ID  Num_Bank_Accounts
ID                                    
0xc7b6   CUS_0x4f2a               -1.0
0xc7b7   CUS_0x4f2a               -1.0
0xc7b8   CUS_0x4f2a               -1.0
0xc7b9   CUS_0x4f2a               -1.0
0xc7ba   CUS_0x4f2a               -1.0
0xc7bb   CUS_0x4f2a               -1.0
0xc7bc   CUS_0x4f2a               -1.0
0xc7bd   CUS_0x4f2a               -1.0
0x115fe  CUS_0xa878               -1.0
0x115ff  CUS_0xa878               -1.0
0x11600  CUS_0xa878               -1.0
0x11601  CUS_0xa878               -1.0
0x11602  CUS_0xa878               -1.0
0x11603  CUS_0xa878               -1.0
0x11604  CUS_0xa878               -1.0
0x11605  CUS_0xa878               -1.0
0x12a9e  CUS_0x43bc               -0.5
0x12a9f  CUS_0x43bc               -0.5
0x12aa0  CUS_0x43bc               -0.5
0x12aa1  CUS_0x43bc               -0.5
0x12aa2  CUS_0x43bc               -0.5
0x12aa3  CUS_0x43bc               -0.5
0x12aa4  CUS_0x43bc               -0.5
0x12aa5  CUS_0x43bc               -0.5
0x15bfa  CUS_0x5993               -0.5
0x15bfb  CUS_0x5993               -0.5
0x15bfc  CUS_0x5993               -0.5
0x15bfd  CUS_0x5993               -0.5
0x15bfe  CUS_0x5993               -0.5
0x15bff  CUS_0x5993               -0.5
0x15c00  CUS_0x5993               -0.5
0x15c01  CUS_0x5993               -0.5

In [136]:
for i in credit_info['Num_Bank_Accounts']:
  if i < 0:
    credit_info['Num_Bank_Accounts'] = credit_info['Num_Bank_Accounts'].replace(i, 0)

Check if changes were done correctly

In [137]:
credit_info['Num_Bank_Accounts'].describe()

count    100000.000000
mean          5.369840
std           2.592452
min           0.000000
25%           3.000000
50%           5.000000
75%           7.000000
max          10.500000
Name: Num_Bank_Accounts, dtype: float64

All done for Num Bank Accounts

# Num Credit Card

In [138]:
credit_info['Num_Credit_Card'].describe()

count    100000.00000
mean         22.47443
std         129.05741
min           0.00000
25%           4.00000
50%           5.00000
75%           7.00000
max        1499.00000
Name: Num_Credit_Card, dtype: float64

There are thousands values that have to be fixed 

Check which stat method use to change those wrong values

In [139]:
credit_info.groupby(['Customer_ID'])['Num_Credit_Card'].agg(['min','max','median','mean',pd.Series.mode])

min   max  median     mean mode
Customer_ID                                 
CUS_0x1000     5     5     5.0    5.000    5
CUS_0x1009     5     5     5.0    5.000    5
CUS_0x100b     4     4     4.0    4.000    4
CUS_0x1011     3     3     3.0    3.000    3
CUS_0x1013     3     3     3.0    3.000    3
...          ...   ...     ...      ...  ...
CUS_0xff3      6  1168     6.0  195.875    6
CUS_0xff4      7     7     7.0    7.000    7
CUS_0xff6      6     6     6.0    6.000    6
CUS_0xffc      8     8     8.0    8.000    8
CUS_0xffd      7     7     7.0    7.000    7

[12500 rows x 5 columns]

Can't use min because there are some 0 values.

Can't use max ore mean because there are over thousands values.


Median and Mode seem to be the same. I will use Median in this case.

In [140]:
credit_info['Num_Credit_Card'] = credit_info.groupby(by=['Customer_ID'])['Num_Credit_Card'].transform('median')

In [141]:
credit_info['Num_Credit_Card'].describe()

count    100000.000000
mean          5.534640
std           2.066834
min           0.500000
25%           4.000000
50%           5.000000
75%           7.000000
max          11.000000
Name: Num_Credit_Card, dtype: float64

All done for Num Credit Card

# Interest Rate

In [142]:
credit_info['Interest_Rate'].describe()

count    100000.000000
mean         72.466040
std         466.422621
min           1.000000
25%           8.000000
50%          13.000000
75%          20.000000
max        5797.000000
Name: Interest_Rate, dtype: float64

There are values that have to be fixed: thousands

Check which stat method use to change those wrong values

In [143]:
credit_info.groupby(['Customer_ID'])['Interest_Rate'].agg(['min','max','median','mean',pd.Series.mode])

min   max  median    mean  mode
Customer_ID                                 
CUS_0x1000    27    27    27.0   27.00    27
CUS_0x1009    17    17    17.0   17.00    17
CUS_0x100b     1     1     1.0    1.00     1
CUS_0x1011    17    17    17.0   17.00    17
CUS_0x1013     6     6     6.0    6.00     6
...          ...   ...     ...     ...   ...
CUS_0xff3      2  3808     2.0  477.75     2
CUS_0xff4     14    14    14.0   14.00    14
CUS_0xff6      1     1     1.0    1.00     1
CUS_0xffc     27    27    27.0   27.00    27
CUS_0xffd     13    13    13.0   13.00    13

[12500 rows x 5 columns]

Can't use min because the Interest Rate could increased on the following month.

Can't use max ore mean because there are over thousands values.

Median and Mode seem to be the same. I will use Median in this case.

In [144]:
credit_info['Interest_Rate'] = credit_info.groupby(by=['Customer_ID'])['Interest_Rate'].transform('median')

In [145]:
credit_info['Interest_Rate'].describe()

count    100000.00000
mean         14.53208
std           8.74133
min           1.00000
25%           7.00000
50%          13.00000
75%          20.00000
max          34.00000
Name: Interest_Rate, dtype: float64

All done for Interest Rate

# Num of Loan

In [146]:
credit_info['Num_of_Loan'].describe()

count     100000
unique       434
top            3
freq       14386
Name: Num_of_Loan, dtype: object

In [147]:
credit_info['Num_of_Loan'].unique()

array(['4', '1', '3', '967', '-100', '0', '0_', '2', '3_', '2_', '7', '5',
       '5_', '6', '8', '8_', '9', '9_', '4_', '7_', '1_', '1464', '6_',
       '622', '352', '472', '1017', '945', '146', '563', '341', '444',
       '720', '1485', '49', '737', '1106', '466', '728', '313', '843',
       '597_', '617', '119', '663', '640', '92_', '1019', '501', '1302',
       '39', '716', '848', '931', '1214', '186', '424', '1001', '1110',
       '1152', '457', '1433', '1187', '52', '1480', '1047', '1035',
       '1347_', '33', '193', '699', '329', '1451', '484', '132', '649',
       '995', '545', '684', '1135', '1094', '1204', '654', '58', '348',
       '614', '1363', '323', '1406', '1348', '430', '153', '1461', '905',
       '1312', '1424', '1154', '95', '1353', '1228', '819', '1006', '795',
       '359', '1209', '590', '696', '1185_', '1465', '911', '1181', '70',
       '816', '1369', '143', '1416', '455', '55', '1096', '1474', '420',
       '1131', '904', '89', '1259', '527', '1241', '449', 

Fix the values that have "_"

In [148]:
credit_info['Num_of_Loan'] = credit_info['Num_of_Loan'].str.strip('_')

Convert to float

In [149]:
credit_info['Num_of_Loan'] = credit_info['Num_of_Loan'].astype('float')

In [150]:
credit_info['Num_of_Loan'].describe()

count    100000.000000
mean          3.009960
std          62.647879
min        -100.000000
25%           1.000000
50%           3.000000
75%           5.000000
max        1496.000000
Name: Num_of_Loan, dtype: float64

Convert negatives to Na

In [151]:
for i in credit_info['Num_of_Loan']:
  if i < 0:
    credit_info['Num_of_Loan'] = credit_info['Num_of_Loan'].replace(i, np.nan)

Check min, max, median, mean, mode for each customer to see which is the best stats to change Na's and 0s

In [152]:
credit_info.groupby(['Customer_ID'])['Num_of_Loan'].agg(['min','max','median','mean',pd.Series.mode])

min     max  median    mean  mode
Customer_ID                                   
CUS_0x1000   2.0     2.0     2.0    2.00   2.0
CUS_0x1009   4.0  1094.0     4.0  140.25   4.0
CUS_0x100b   0.0     0.0     0.0    0.00   0.0
CUS_0x1011   3.0     3.0     3.0    3.00   3.0
CUS_0x1013   3.0     3.0     3.0    3.00   3.0
...          ...     ...     ...     ...   ...
CUS_0xff3    3.0     3.0     3.0    3.00   3.0
CUS_0xff4    5.0     5.0     5.0    5.00   5.0
CUS_0xff6    2.0     2.0     2.0    2.00   2.0
CUS_0xffc    8.0     8.0     8.0    8.00   8.0
CUS_0xffd    6.0     6.0     6.0    6.00   6.0

[12500 rows x 5 columns]

I will use Mode for Na's and Median for the thousands values

In [153]:
credit_info['Num_of_Loan'] = credit_info.groupby('Customer_ID')['Num_of_Loan'].transform(lambda x: x.fillna(stats.mode(x)[0][0]))

In [154]:
credit_info['Num_of_Loan'] = credit_info.groupby(by=['Customer_ID'])['Num_of_Loan'].transform('median')

In [155]:
credit_info['Num_of_Loan'].describe()

count    100000.000000
mean          3.532880
std           2.446356
min           0.000000
25%           2.000000
50%           3.000000
75%           5.000000
max           9.000000
Name: Num_of_Loan, dtype: float64

All done with Num_of_Loan

# Type of Loan

In [156]:
credit_info['Type_of_Loan'].describe()

count             88592
unique             6260
top       Not Specified
freq               1408
Name: Type_of_Loan, dtype: object

In [157]:
credit_info['Type_of_Loan'].unique()

array(['Auto Loan, Credit-Builder Loan, Personal Loan, and Home Equity Loan',
       'Credit-Builder Loan', 'Auto Loan, Auto Loan, and Not Specified',
       ..., 'Home Equity Loan, Auto Loan, Auto Loan, and Auto Loan',
       'Payday Loan, Student Loan, Mortgage Loan, and Not Specified',
       'Personal Loan, Auto Loan, Mortgage Loan, Student Loan, and Student Loan'],
      dtype=object)

In [158]:
credit_info['Type_of_Loan'].isna().sum()

11408

Fill Na's with 'Not Specified'

In [159]:
credit_info['Type_of_Loan'] = credit_info['Type_of_Loan'].fillna('Not Specified')

In [160]:
credit_info['Type_of_Loan'].isna().sum()

0

Convert to list values

In [161]:
credit_info_test = credit_info

In [162]:
#for i in credit_info_test['Type_of_Loan']:
 # replacement = map(lambda x:x.strip(), i.replace('and','').split(','))
  #credit_info_test['Type_of_Loan'] = credit_info_test['Type_of_Loan'].replace(i, replacement)

In [163]:
# Not sure how to change each value to list value

# Delay from due date

In [164]:
credit_info['Delay_from_due_date'].describe()

count    100000.000000
mean         21.068780
std          14.860104
min          -5.000000
25%          10.000000
50%          18.000000
75%          28.000000
max          67.000000
Name: Delay_from_due_date, dtype: float64

In [165]:
# Negatives values could mean as number of days paid in advance
# Not sure if I have to delete them

In [166]:
credit_info.groupby(['Customer_ID'])['Delay_from_due_date'].agg(['min','max','median','mean',pd.Series.mode])

min  max  median    mean mode
Customer_ID                               
CUS_0x1000    57   67    62.0  62.250   62
CUS_0x1009     3   10     8.0   7.250    8
CUS_0x100b     9   19    14.0  13.500   14
CUS_0x1011    27   29    27.0  27.250   27
CUS_0x1013    12   16    12.0  12.500   12
...          ...  ...     ...     ...  ...
CUS_0xff3      9   14    13.0  12.250   13
CUS_0xff4     15   16    16.0  15.875   16
CUS_0xff6     -4    1     1.0  -0.125    1
CUS_0xffc     41   46    46.0  45.375   46
CUS_0xffd     19   25    23.0  23.000   23

[12500 rows x 5 columns]

In [167]:
# Diferent values for each customer
# Change the values to the median or leave it because each entry of each customer is in a different month?

There are different values for each customer. 

I will use the median to change the values. 

In [168]:
# credit_info['Delay_from_due_date'] = credit_info.groupby(by=['Customer_ID'])['Delay_from_due_date'].transform('median')

In [169]:
# credit_info[['Customer_ID','Delay_from_due_date']][credit_info['Delay_from_due_date'] < 0]

In [170]:
# print(credit_info[['Customer_ID']][credit_info['Delay_from_due_date'] < 0].count())
# print()
# credit_info[['Customer_ID']][credit_info['Delay_from_due_date'] < 0].groupby('Customer_ID').Customer_ID.count() 

# Num of Delayed Payment

In [171]:
credit_info['Num_of_Delayed_Payment'].describe()

count     92998
unique      749
top          19
freq       5327
Name: Num_of_Delayed_Payment, dtype: object

In [172]:
credit_info['Num_of_Delayed_Payment'].unique()

array(['7', nan, '4', '8_', '6', '1', '-1', '3_', '0', '8', '5', '3', '9',
       '12', '15', '17', '10', '2', '2_', '11', '14', '20', '22', '13',
       '13_', '14_', '16', '12_', '18', '19', '23', '24', '21', '3318',
       '3083', '22_', '1338', '4_', '26', '11_', '3104', '21_', '25',
       '10_', '183_', '9_', '1106', '834', '19_', '24_', '17_', '23_',
       '2672', '20_', '2008', '-3', '538', '6_', '1_', '16_', '27', '-2',
       '3478', '2420', '15_', '707', '708', '26_', '18_', '3815', '28',
       '5_', '1867', '2250', '1463', '25_', '7_', '4126', '2882', '1941',
       '2655', '2628', '132', '3069', '306', '0_', '3539', '3684', '1823',
       '4128', '1946', '827', '2297', '2566', '904', '182', '929', '3568',
       '2503', '1552', '2812', '1697', '3764', '851', '3905', '923', '88',
       '1668', '3253', '808', '2689', '3858', '642', '3457', '1402',
       '1732', '3154', '847', '3037', '2204', '3103', '1063', '2056',
       '1282', '1841', '2569_', '211', '793', '3484', '4

In [173]:
credit_info[['Customer_ID','Num_of_Delayed_Payment']].head(24)

Customer_ID Num_of_Delayed_Payment
ID                                       
0x1602   CUS_0xd40                      7
0x1603   CUS_0xd40                    NaN
0x1604   CUS_0xd40                      7
0x1605   CUS_0xd40                      4
0x1606   CUS_0xd40                    NaN
0x1607   CUS_0xd40                      4
0x1608   CUS_0xd40                     8_
0x1609   CUS_0xd40                      6
0x160e  CUS_0x21b1                      4
0x160f  CUS_0x21b1                      1
0x1610  CUS_0x21b1                     -1
0x1611  CUS_0x21b1                     3_
0x1612  CUS_0x21b1                      1
0x1613  CUS_0x21b1                      0
0x1614  CUS_0x21b1                      4
0x1615  CUS_0x21b1                      4
0x161a  CUS_0x2dbc                      8
0x161b  CUS_0x2dbc                      6
0x161c  CUS_0x2dbc                      7
0x161d  CUS_0x2dbc                      5
0x161e  CUS_0x2dbc                      5
0x161f  CUS_0x2dbc                      6
0x1620  CUS_0x2dbc                      6
0x1621  CUS_0x2dbc                      6

There are "_",  negatives, and Na values

First: fix the values that have "_"

In [174]:
credit_info['Num_of_Delayed_Payment'] = credit_info['Num_of_Delayed_Payment'].str.strip('_')

Second: change negatives values to Na's

In [176]:
credit_info['Num_of_Delayed_Payment'] = credit_info['Num_of_Delayed_Payment'].astype('float')

In [177]:
for i in credit_info['Num_of_Delayed_Payment']:
  if i < 0:
    credit_info['Num_of_Delayed_Payment'] = credit_info['Num_of_Delayed_Payment'].replace(i, np.nan)

Third: change Na values

In [178]:
credit_info.groupby(['Customer_ID'])['Num_of_Delayed_Payment'].agg(['min','max','median','mean',pd.Series.mode])

min     max  median        mean          mode
Customer_ID                                                
CUS_0x1000   23.0    28.0    25.0   25.000000          25.0
CUS_0x1009   16.0  1749.0    18.0  234.250000          18.0
CUS_0x100b    7.0     9.0     7.0    7.375000           7.0
CUS_0x1011   12.0    17.0    15.0   14.500000          15.0
CUS_0x1013    7.0     9.0     9.0    8.428571           9.0
...           ...     ...     ...         ...           ...
CUS_0xff3     6.0    11.0     9.0    8.625000           9.0
CUS_0xff4    10.0    14.0    13.0   12.125000  [10.0, 13.0]
CUS_0xff6     2.0     7.0     4.0    4.000000           4.0
CUS_0xffc    14.0    19.0    16.0   16.250000          16.0
CUS_0xffd     9.0    12.0    12.0   11.428571          12.0

[12500 rows x 5 columns]

I will use Median. There are Mode values that are lists.

In [179]:
credit_info['Num_of_Delayed_Payment'] = credit_info.groupby(by=['Customer_ID'])['Num_of_Delayed_Payment'].transform('median')

In [180]:
credit_info['Num_of_Delayed_Payment'].describe()

count    100000.000000
mean         13.320960
std           6.182172
min           0.000000
25%           9.000000
50%          14.000000
75%          18.000000
max          26.500000
Name: Num_of_Delayed_Payment, dtype: float64

All done with Num of Delayed Payment

# Changed Credit Limit

In [181]:
credit_info['Changed_Credit_Limit'].describe()

count     100000
unique      3635
top            _
freq        2091
Name: Changed_Credit_Limit, dtype: object

In [182]:
credit_info['Changed_Credit_Limit'].unique()

array(['11.27', '_', '6.27', ..., '27.38', '25.16', '21.17'], dtype=object)

Delete "_" values

In [183]:
credit_info['Changed_Credit_Limit'] = credit_info['Changed_Credit_Limit'].str.strip('_')

In [184]:
credit_info['Changed_Credit_Limit'].unique()

array(['11.27', '', '6.27', ..., '27.38', '25.16', '21.17'], dtype=object)

Change empty values for Nas

In [185]:
for i in credit_info['Changed_Credit_Limit']:
  if i == '':
    credit_info['Changed_Credit_Limit'] = credit_info['Changed_Credit_Limit'].replace(i, np.nan)

Convert Changed_Credit_Limit type to float

In [186]:
credit_info['Changed_Credit_Limit'] = credit_info['Changed_Credit_Limit'].astype('float')

In [187]:
credit_info['Changed_Credit_Limit'].describe()

count    97909.000000
mean        10.389025
std          6.789496
min         -6.490000
25%          5.320000
50%          9.400000
75%         14.870000
max         36.970000
Name: Changed_Credit_Limit, dtype: float64

Change negatives for Nas

In [188]:
for i in credit_info['Changed_Credit_Limit']:
  if i < 0:
    credit_info['Changed_Credit_Limit'] = credit_info['Changed_Credit_Limit'].replace(i, np.nan)

Check which stat use to replace the Nas

In [189]:
credit_info.groupby(['Customer_ID'])['Changed_Credit_Limit'].agg(['min','max','median','mean',pd.Series.mode])

min    max  median       mean   mode
Customer_ID                                        
CUS_0x1000    1.63   2.63    1.63   1.880000   1.63
CUS_0x1009    9.73   9.73    9.73   9.730000   9.73
CUS_0x100b    8.34  11.34   11.34  10.911429  11.34
CUS_0x1011   13.42  14.42   14.42  14.170000  14.42
CUS_0x1013    1.33   4.33    1.33   1.705000   1.33
...            ...    ...     ...        ...    ...
CUS_0xff3     6.86  13.86    6.86   8.485000   6.86
CUS_0xff4     7.83  10.83    7.83   8.205000   7.83
CUS_0xff6    10.40  16.40   10.40  11.900000   10.4
CUS_0xffc     5.82  12.82    8.82   8.820000   8.82
CUS_0xffd    10.07  12.07   10.07  10.320000  10.07

[12500 rows x 5 columns]

Change Nas with Mode

In [190]:
credit_info['Changed_Credit_Limit'] = credit_info.groupby('Customer_ID')['Changed_Credit_Limit'].transform(lambda x: x.fillna(stats.mode(x)[0][0]))

Check if all Nas where fixed

In [191]:
credit_info['Changed_Credit_Limit'].describe()

count    99982.000000
mean        10.473497
std          6.657536
min          0.000000
25%          5.370000
50%          9.400000
75%         14.850000
max         36.970000
Name: Changed_Credit_Limit, dtype: float64

In [192]:
credit_info['Changed_Credit_Limit'].isna().sum()

18

There 18 that have not been fixed because all their values were negatives before converting them to Na.

I will change them to 0

In [193]:
credit_info['Changed_Credit_Limit'] = credit_info['Changed_Credit_Limit'].fillna(0)

Check if all Nas where fixed

In [194]:
credit_info['Changed_Credit_Limit'].describe()

count    100000.000000
mean         10.471612
std           6.658419
min           0.000000
25%           5.370000
50%           9.400000
75%          14.850000
max          36.970000
Name: Changed_Credit_Limit, dtype: float64

All done with Changed Credit Limit

# Num Credit Inquiries

In [195]:
credit_info['Num_Credit_Inquiries'].describe()

count    98035.000000
mean        27.754251
std        193.177339
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max       2597.000000
Name: Num_Credit_Inquiries, dtype: float64

There are outliers

Check which stat use to replace the outliers

In [196]:
credit_info.groupby(['Customer_ID'])['Num_Credit_Inquiries'].agg(['min','max','median','mean',pd.Series.mode])

min     max  median        mean        mode
Customer_ID                                              
CUS_0x1000   10.0    11.0    11.0   10.875000        11.0
CUS_0x1009    2.0     4.0     3.0    3.000000  [2.0, 4.0]
CUS_0x100b    1.0  2271.0     1.0  285.500000         1.0
CUS_0x1011    7.0  1965.0     7.0  251.750000         7.0
CUS_0x1013    3.0     3.0     3.0    3.000000         3.0
...           ...     ...     ...         ...         ...
CUS_0xff3     3.0     5.0     5.0    4.250000         5.0
CUS_0xff4     5.0     5.0     5.0    5.000000         5.0
CUS_0xff6     2.0     2.0     2.0    2.000000         2.0
CUS_0xffc     8.0    13.0    13.0   12.375000        13.0
CUS_0xffd     7.0  1801.0     7.0  263.285714         7.0

[12500 rows x 5 columns]

I will use Median

In [197]:
credit_info['Num_Credit_Inquiries'] = credit_info.groupby('Customer_ID')['Num_Credit_Inquiries'].transform('median')

Check if issues were fixed

In [198]:
credit_info['Num_Credit_Inquiries'].describe()

count    100000.000000
mean          5.794600
std           3.831618
min           0.000000
25%           3.000000
50%           5.000000
75%           8.500000
max          17.000000
Name: Num_Credit_Inquiries, dtype: float64

All done with Num Credit Inquiries

# Credit Mix

In [199]:
credit_info['Credit_Mix'].describe()

count       100000
unique           4
top       Standard
freq         36479
Name: Credit_Mix, dtype: object

In [200]:
credit_info['Credit_Mix'].unique()

array(['_', 'Good', 'Standard', 'Bad'], dtype=object)

In [201]:
credit_info['Credit_Mix'][credit_info['Credit_Mix'] == '_'].count()

20195

There a incorrect value '_'

I am going to change it to Na

In [202]:
for i in credit_info['Credit_Mix']:
  if i == '_':
    credit_info['Credit_Mix'] = credit_info['Credit_Mix'].replace(i, np.nan)

Now I will replace it for the Mode

In [203]:
credit_info_test['Credit_Mix'] = credit_info_test.groupby('Customer_ID')['Credit_Mix'].transform(lambda x: x.fillna(stats.mode(x)[0][0]))

In [204]:
credit_info['Credit_Mix'].unique()

array(['Good', 'Standard', 'Bad'], dtype=object)

In [205]:
credit_info['Credit_Mix'].isna().sum()

0

All done with Credit Mix